### PR TITLE
Add photos to contacts

### PR DESCRIPTION
**Pull request by nashirj**

### CODE REVIEW

**Comment by mplekunov** [comment]
chmod +rw on the folder where images will be stored doesn't give enough permissions to "load" them from the server, giving full permissions (777) solved an issue.

**Comment by mplekunov** [comment]
Image size seem to have an effect on whether it will throw an error or not, I used 117 Kb image and 2 Kb image. 117 wasn't loaded when I tried to search for a contact because "the setItem cahcedContacts exceeded quota". From my research, this happens when localStorage is "full". My guess is, the size of the object that can be stored in localStorage is limited and it throws an error when you exceed that limit.

**Comment by mplekunov** [comment]
Fixed doDeleteContact URI, it was sending delete request to the update endpoint

**Comment by mplekunov** [comment]
> Image size seem to have an effect on whether it will throw an error or not, I used 117 Kb image and 2 Kb image. 117 wasn't loaded when I tried to search for a contact because "the setItem cahcedContacts exceeded quota". From my research, this happens when localStorage is "full". My guess is, the size of the object that can be stored in localStorage is limited and it throws an error when you exceed that limit.

After some additional investigation it looks like when it saves image on the server it increases it's size DRASTICALLY... The size of the saved image is not 117 Kb but 5 Mb

https://i.imgur.com/KZLdqfV.jpeg

File: js/contact.js [hunk 1]
```diff
@@ -2,7 +2,7 @@ async function doDeleteContact() {
   if (confirm("Are you sure you want to delete this contact?")) {
     const contact = JSON.parse(localStorage.getItem('individualContact'));
     const [status, responseJson] = await postData(
```
Comment (mplekunov) [inline js/contact.js:4]:
```suggestion
    const [status, responseJson] = await postData(
```
For the Delete function you will have to create another await helper function "deleteData" which will have request method set to "DELETE"

**Comment by mplekunov** [comment]
I fixed DELETE call to the DeleteContact endpoint, now it does what it supposed to do, I also tested it so it works.

**Comment by mplekunov** [comment]
> > Image size seem to have an effect on whether it will throw an error or not, I used 117 Kb image and 2 Kb image. 117 wasn't loaded when I tried to search for a contact because "the setItem cahcedContacts exceeded quota". From my research, this happens when localStorage is "full". My guess is, the size of the object that can be stored in localStorage is limited and it throws an error when you exceed that limit.
> 
> After some additional investigation it looks like when it saves image on the server it increases it's size DRASTICALLY... The size of the saved image is not 117 Kb but 5 Mb. Since most browsers have a default localStorage set to 5 MB it's natural for it to throw an exception... Could we, perhaps, use some other space to store images downloaded by the page? Maybe in temp folder on the local storage of the machine that opened page ? I can also try and do compression of the image on the backend side, but from what I read, storing images in localstorage is no good in any case.
> 
> 

After a bit of testing with compression, changing compression to 1 (it can vary from 0 to 9 where 0 is no compression and 9 is the best compression) reduces saved image size to 137 Kb (from 5 Mb we had in 0 compression). So, it looks like I will have to use compression in order to save some space. That being said, I think we still need to find some other place where we will be able to store images because if we have ~37 contacts with 137 Kb images, we will run out of space in localStorage

**Comment by mplekunov** [comment]
now you should have up-to-date changes to compression on your branch

**Comment by nashirj** [comment]
Thanks for all the comments. As far as directory permissions, I [read](https://stackoverflow.com/questions/5357781/giving-php-write-permission-in-apache) that we should use `sudo chown www-data <imagedir>` to give apache permissions on the image directory.

As far as updates to handle the large image sizes, here's what we discussed over discord for others to see:
1) Two endpoints for contacts
- Get all contacts matching query (just returns first name, last name, id); this is used for home page.
- Get contact by ID (returns all contact info including base64 image); this is used on the individual contact page to display email, phone number, and image
2) Store the image ID in the contact, and return that to the frontend; the frontend then has to request an ID for a contact if it wants to display the image for that contact. so a user would get all contacts matching search string (which would only contain image id) and then when they go to an individual contact page, would do a new get request for the image.

@mplekunov prefers option 2, so I think we are leaning towards that.

I'm also going to add image resizing to the JS before the image is uploaded to mitigate some of the space concerns.

**Pull request merged**